cocotb/run.py

```diff
@@ -0,0 +1,929 @@
+from distutils.sysconfig import get_config_var
+
+_magic_re = re.compile(r"([\\{}])")
+_space_re = re.compile(r"([\s])", re.ASCII)
+
+
+def as_tcl_value(value):
+    # add '\' before special characters and spaces
+    value = _magic_re.sub(r"\\\1", value)
+    value = value.replace("\n", r"\n")
+    value = _space_re.sub(r"\\\1", value)
+    if value[0] == '"':
+        value = "\\" + value
+
+    return value
+
+
+class Simulator(object):
+    def __init__(
+        self,
+        toplevel,
+        module,
+        work_dir=None,
+        python_search=None,
+        toplevel_lang="verilog",
+        verilog_sources=None,
+        vhdl_sources=None,
+        includes=None,
+        defines=None,
+        parameters=None,
+        compile_args=None,
+        sim_args=None,
+        extra_args=None,
+        plus_args=None,
+        force_compile=False,
+        testcase=None,
+        sim_build="sim_build",
+        seed=None,
+        extra_env=None,
+        compile_only=False,
+        waves=None,
+        gui=False
+    ):
+
+        self.logger = logging.getLogger()
+
+        self.sim_dir = os.path.abspath(sim_build)
+        os.makedirs(self.sim_dir, exist_ok=True)
+
+        self.lib_dir = os.path.join(os.path.dirname(cocotb.__file__), "libs")
+
+        self.lib_ext = "so"
+        if os.name == "nt":
+            self.lib_ext = "dll"
+
+        if isinstance(module, str):
+            self.module = module
+        else:
+            self.module = ",".join(module)
+
+        self.work_dir = self.sim_dir
+
+        if work_dir is not None:
+            absworkdir = os.path.abspath(work_dir)
+            if os.path.isdir(absworkdir):
+                self.work_dir = absworkdir
+
+        if python_search is None:
+            python_search = []
+
+        self.python_search = python_search
+
+        self.toplevel = toplevel
+        self.toplevel_lang = toplevel_lang
+
+        if verilog_sources is None:
+            verilog_sources = []
+
+        self.verilog_sources = self.get_abs_paths(verilog_sources)
+
+        if vhdl_sources is None:
+            vhdl_sources = []
+
+        self.vhdl_sources = self.get_abs_paths(vhdl_sources)
+
+        if includes is None:
+            includes = []
+
+        self.includes = self.get_abs_paths(includes)
+
+        if defines is None:
+            defines = []
+
+        self.defines = defines
+
+        if parameters is None:
+            parameters = {}
+
+        self.parameters = parameters
+
+        if compile_args is None:
+            compile_args = []
+
+        if extra_args is None:
+            extra_args = []
+
+        self.compile_args = compile_args + extra_args
+
+        if sim_args is None:
+            sim_args = []
+
+        self.sim_args = sim_args + extra_args
+
+        if plus_args is None:
+            plus_args = []
+
+        self.plus_args = plus_args
+        self.force_compile = force_compile
+        self.compile_only = compile_only
+
```

Comment:
This avoidance of mutable defaults creates huge additional verbosity and reduces the effectiveness of Sphinx / `help` - I'd usually recommend just making sure to copy all the argument before there's any chance of anyone mutating them:

```suggestion
    def __init__(
        self,
        toplevel,
        module,
        work_dir=None,
        python_search=[],
        toplevel_lang="verilog",
        verilog_sources=[],
        vhdl_sources=[],
        includes=[],
        defines=[],
        parameters={},
        compile_args=[],
        sim_args=[],
        extra_args=[],
        plus_args=[],
        force_compile=False,
        testcase=None,
        sim_build="sim_build",
        seed=None,
        extra_env=None,
        compile_only=False,
        waves=None,
        gui=False
    ):

        self.logger = logging.getLogger()

        self.sim_dir = os.path.abspath(sim_build)
        os.makedirs(self.sim_dir, exist_ok=True)

        self.lib_dir = os.path.join(os.path.dirname(cocotb.__file__), "libs")

        self.lib_ext = "so"
        if os.name == "nt":
            self.lib_ext = "dll"

        if isinstance(module, str):
            self.module = module
        else:
            self.module = ",".join(module)

        self.work_dir = self.sim_dir

        if work_dir is not None:
            absworkdir = os.path.abspath(work_dir)
            if os.path.isdir(absworkdir):
                self.work_dir = absworkdir

        # note: to avoid mutating argument defaults, we ensure that no value
        # is written without a copy. This is much more concise and leads to
        # a better docstring than using `None` as a default in the parameters
        # list.
        self.python_search = list(python_search)
        self.toplevel = toplevel
        self.toplevel_lang = toplevel_lang
        self.verilog_sources = self.get_abs_paths(verilog_sources)
        self.vhdl_sources = self.get_abs_paths(vhdl_sources)
        self.includes = self.get_abs_paths(includes)
        self.defines = list(defines)
        self.parameters = dict(parameters)
        self.compile_args = compile_args + extra_args
        self.sim_args = sim_args + extra_args
        self.plus_args = list(plus_args)
        self.force_compile = force_compile
        self.compile_only = compile_only

```

Reply:
`None` is really nice in that wrappers don't have to be aware of, and replicate, the default value. Really, I think this should probably be a configuration dataclass. We can replicate some of the functionality in that dataclass now, then remove it and just use `@dataclass` once we drop Python 3.6 support (it's EOL at the end of this year).

Reply:
True, although if they just forward `**kwargs` then they don't need to be aware of the argument at all. No objection to using data classes, although I have no experience with them either.